jasmine-standalone-2.3.4 (4)/src/Game.js

```diff
@@ -0,0 +1,156 @@
+function Game() {
+  this.frames = [];
+  this.pairs = [];
+}
+
+ // GENERAL METHODS
+
+Game.prototype.isInProgress = function() {
+  return (this._endGameStrike() || this._endGameSpare() || this._lessThan20());
+}
+
+Game.prototype._lastFrame = function() {
+  return this.frames[this.frames.length - 1];
+}
+
+Game.prototype._secondLastFrame = function() {
+  return this.frames[this.frames.length - 2];
+}
+
+Game.prototype._numberOfPairs = function() {
+  return this.pairs.length;
+}
+
+Game.prototype._valueOfPair = function(i) {
+  return Object.values(this.pairs[i])[0];
+}
+
+// ADD FRAME
+
+Game.prototype._addFrame = function() {
+  let framesCount = this.frames.length;
+  let lastFrame = this._lastFrame();
+  let secondLastFrame = this._secondLastFrame();
+
+  if (framesCount === 0 || framesCount === 1) {
+    this.frames.push(1);
+  } else if (lastFrame !== secondLastFrame) {
+    this.frames.push(lastFrame);
+  } else if (lastFrame === secondLastFrame) {
+    if (lastFrame === 10) {
+      this.frames.push(10);
+    } else {
+      this.frames.push(lastFrame + 1);
+    }
+  }
+
+}
+
+// COUNTING METHODS
+
+Game.prototype.basicScore = function() {
+  var counter = 0;
+  for (i = 0; i < this.pairs.length; i ++) {
+   if (this._valueOfPair(i) !== undefined) {
+      if (i < 18) {
+        counter += this._valueOfPair(i)
+      } else if (i === 18 && this._valueOfPair(18) === 10){
+        counter += this._valueOfPair(18)
+      } else if  (i === 18 && this._valueOfPair(18) !== 10 ) {
+        counter += this._valueOfPair(18)
+        counter += this._valueOfPair(19)
+      }
+    }
+  }
+  return counter;
+}
+
+Game.prototype.spares = function() {
+  var spares = 0;
+  for (i = 0; i < 17; i += 2) {
+    if (!this.isStrike(this._valueOfPair(i))) { // if strike - cannot be spare
+      if (this._valueOfPair(i) + this._valueOfPair(i + 1) === 10) {
+        spares += this._valueOfPair(i + 2);
+      }
+    }
+  }
+  if (this._valueOfPair(18) + this._valueOfPair(19) === 10) {
+    spares += this._valueOfPair(20)
+  }
+  return spares;
+}
+
+Game.prototype.strikes = function() {
+  var strikes = 0;
+  for (i = 0; i <= 14; i += 2) {
```

Comment:
I don't really understand what is going on here. I think that we're trying to get the bonuses from the next frame in the case of a strike. When performing calculations, it's worth putting them in strongly named methods so it is easy for another developer to read what is happening. Eg:

```
if (!nextFrameIsAStrike()) {
  strikes += getRollsFromNextFrame()
} else {
  strikes += getFirstRollFromNextFrameAndFirstRollFromFrameAfter()
}
```

Obviously that last method name is really long and feels kind of unnecessary... which is another reason to do this. Because when you end up with method names like that, it's another hint that you should probably refactor your code (see comment above). If I can't express the logic in a simple, well named method, then I probably need to interrogate it.

For example, here I would be thinking 'actually, really what I want is just the next two rolls... it's just that if the next frame is a strike, then it has only a single roll. How could I refactor this so that didn't matter?'.